### PR TITLE
Consistency updates

### PR DESCRIPTION
**Pull request by mj-will**
Initial changes to make the search and PE codes consistent.

**Changes**

- Replace references to early warning with pre-merger

`generate_data_lisa_ew` -> `generate_data_lisa_pre_merger`

- Fix additional end data using `cyclic_time_shift`
- Zero data after final ifft (forward zeroes and around cutoff time)
- Set seed to `137`

`generate_waveform_lisa_ew` -> `generate_waveform_lisa_pre_merger`

- Zero data after final ifft (forward zeroes and around cutoff time)
- Use common function `apply_pre_merger_kernel` to avoid duplicate code

`generate_early_warning_psds` -> `generate_pre_merger_psds`

- Remove `tlen` and use `duration` instead
- General formatting
- Function simplified to only generate one PSD at a time. This avoids duplicate code

`waveform.py`

- Remove unused `FDomainLISAEWGenerator`

### CODE REVIEW

File: pycbc/psd/lisa_pre_merger.py [hunk 1]
```diff
@@ -364,234 +362,72 @@ def linear_phase_fir_kernel_to_minimum_phase_whitening_fir_kernel(
         #
 
         return kernel, phase
+    
 
-def generate_early_warning_psds(
+def generate_pre_merger_psds(
```
Comment (GarethCabournDavies) [inline pycbc/psd/lisa_pre_merger.py:367]: I have confirmed that the outputs of this function match the utils.py version in `lisa_ew_detection`

Comment (GarethCabournDavies) [reply pycbc/psd/lisa_pre_merger.py:367]: I also didnt realise i was on main, so using this function is now part of that repo! [this commit](https://github.com/icg-gravwaves/lisa_ew_detection/commit/155222f1836ebf1cae9985cc536b85e0fb30401a)

File: pycbc/psd/lisa_pre_merger.py [hunk 2]
```diff
@@ -90,8 +88,8 @@ def psd_to_linear_phase_whitening_fir_kernel(
                 the Nyquist frequency.
 
         Returns:
-            Tuple[numpy.ndarray, int, int], the kernel, latency,
-            sample rate pair.  The kernel is a numpy array containing
+            Tuple[np.ndarray, int, int], the kernel, latency,
```
Comment (GarethCabournDavies) [inline pycbc/psd/lisa_pre_merger.py:91]: docstrings should use `numpy`, not `np`

**Comment by GarethCabournDavies** [comment]
Note that the MR for utilising these changes is on https://github.com/icg-gravwaves/lisa_ew_detection/pull/11

File: pycbc/waveform/pre_merger_waveform.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        uid=uids[1],
+    )
+    # Zero initial data
+    tout_ww.data[:nefz] = 0
+    # Zero from cutoff
+    tout_ww.data[-nctf:] = 0
+    return tout_ww
+
+
+def generate_data_lisa_pre_merger(
```
Comment (GarethCabournDavies) [inline pycbc/waveform/pre_merger_waveform.py:71]: I have confirmed that the latest version of this (7001413) matches the output of the function from utils.py

File: pycbc/waveform/pre_merger_waveform.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    psds_for_whitening: dict[str: FrequencySeries]
+        Power spectral denisities for whitening in the frequency-domain.
+    sample_rate : float
+        Sampling rate.
+    window_length : int
+        Length (in samples) of time-domain window applied to the start of the
+        waveform.
+    cutoff_time: float
+        Time (in seconds) from the end of the waveform to cutoff.
+    kernel_length : int
```
Comment (GarethCabournDavies) [inline pycbc/waveform/pre_merger_waveform.py:219]: Can this be removed then?

File: pycbc/waveform/pre_merger_waveform.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+_WINDOW = None
+def generate_waveform_lisa_pre_merger(
+    waveform_params,
+    psds_for_whitening,
+    sample_rate,
+    window_length,
+    cutoff_time,
+    kernel_length,
+    extra_forward_zeroes=0,
+):
```
Comment (GarethCabournDavies) [inline pycbc/waveform/pre_merger_waveform.py:200]: I've also confirmed that this matches the utils version (after the necessary tweaks to the utils version to do the extra zeroing step)

File: pycbc/waveform/pre_merger_waveform.py (hunk 4 above)
Comment (mj-will) [reply pycbc/waveform/pre_merger_waveform.py:219]: It probably can, we can always add it again if need be.

File: pycbc/psd/lisa_pre_merger.py [hunk 6]
```diff
@@ -364,234 +362,72 @@ def linear_phase_fir_kernel_to_minimum_phase_whitening_fir_kernel(
...
-        generator.
-    psds_for_whitening: dict[str: FrequencySeries]
-        Power spectral denisities for whitening in the frequency-domain.
-    window_length : int
-        Length (in samples) of time-domain window applied to the start of the
-        waveform.
-    cutoff_time: float
-        Time (in seconds) from the end of the waveform to cutoff.
+    psd_file : str
+        Path to the PSD file. Assumes the same PSD for A and E channels.
```
Comment (GarethCabournDavies) [inline pycbc/psd/lisa_pre_merger.py:378]: I don't think this assumption is made in the function

Comment (mj-will) [reply pycbc/psd/lisa_pre_merger.py:378]: Indeed, I suspect this is a holdover from the old function

File: pycbc/psd/lisa_pre_merger.py (hunk 2 above)
Comment (mj-will) [reply pycbc/psd/lisa_pre_merger.py]: I've addressed and updated the other doc-strings for this class.

File: pycbc/psd/lisa_pre_merger.py (hunk 6 above)
Comment (mj-will) [reply pycbc/psd/lisa_pre_merger.py]: Resolved in https://github.com/icg-gravwaves/pycbc/pull/1/commits/7cf886b23fadae3bfed85b11a432dcb175e6d068

File: pycbc/waveform/pre_merger_waveform.py (hunk 4 above)
Comment (mj-will) [reply pycbc/waveform/pre_merger_waveform.py:218]: I've removed this in https://github.com/icg-gravwaves/pycbc/pull/1/commits/050675e33817e74b9509d2ce861841032001e0b8

**Comment by mj-will** [comment]
We will likely need the change in https://github.com/gwastro/pycbc/pull/4766/files for these changes to work.

**Review by GarethCabournDavies — APPROVED** [review]

**Review by spxiwh — APPROVED** [review]
I'm happy if you've both checked over and verified all of this!

**Comment by mj-will** [comment]
I'll need to put in the fix when discussed today Ian, but I think I'll do that in a separate MR now.

**Pull request merged**